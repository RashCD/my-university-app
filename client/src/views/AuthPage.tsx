import { navigate, RouteComponentProps } from '@reach/router';
import React, { useContext, useState } from 'react';
import AppHeader from '../components/AppHeader';
import Button from '../components/Button';
import Styles from '../assets/styles/views/AuthPage.module.scss';
import LoginForm from '../components/LoginForm';
import SignupForm from '../components/SignupForm';
import requestLogin from '../queries/requestLogin';
import requestSignup from '../queries/requestSignup';
import { UserContext } from '../context/UserContext';

enum AuthType {
  LOGIN = 'login',
  SIGNUP = 'signup',
}

const AuthPage = (props: RouteComponentProps) => {
  const { setLoggedIn } = useContext(UserContext);
  const [buttonSelected, setButtonSelected] = useState(AuthType.LOGIN);

  const loggedInCallback = () => {
    setLoggedIn(true);
    navigate('/profile', { replace: true });
  };

  return (
    <div className="app">
      <AppHeader />
      <main className="app-main">
        <div className={Styles.container}>
          <Button
            className={
              buttonSelected !== AuthType.LOGIN && Styles.selectedRight
            }
            onButtonClick={() => setButtonSelected(AuthType.LOGIN)}
          >
            Login
          </Button>
          <Button
            className={
              buttonSelected !== AuthType.SIGNUP && Styles.selectedLeft
            }
            onButtonClick={() => setButtonSelected(AuthType.SIGNUP)}
          >
            Sign Up
          </Button>
          <div className={Styles.formContainer}>
            {buttonSelected === AuthType.LOGIN ? (
              <LoginForm
                onFormSubmit={(event) =>
                  requestLogin(event)
                    .then((res) => res.status === 200)
                    .then(loggedInCallback)
                    .catch((err) => alert(err.response.data.message))
                }
              />
            ) : (
              <SignupForm
                onFormSubmit={(event) =>
                  requestSignup(event)
                    .then((res) => res.status === 200)
                    .then(loggedInCallback)
                    .catch((err) => alert(err.response.data.message))
                }
              />
            )}
          </div>
        </div>
      </main>
    </div>
  );
};

export default AuthPage;
