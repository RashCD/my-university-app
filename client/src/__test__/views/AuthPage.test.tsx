import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';
import {
  render,
  cleanup,
  screen,
  fireEvent,
  act,
} from '@testing-library/react';
import AuthPage from '../../views/AuthPage';

afterEach(cleanup);

function renderWithRouter(
  ui: JSX.Element,
  { route = '/', history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
}

describe('Test for auth page', () => {
  beforeEach(() => {
    renderWithRouter(<AuthPage />);
  });

  test('should have login tab & login section', () => {
    const loginButtons = screen.getAllByText('Login');

    const loginButton = loginButtons.find(
      (button) => button.className === 'button'
    );

    expect(screen.getByText('Username')).toBeInTheDocument();

    expect(screen.getByText('Password')).toBeInTheDocument();

    expect(loginButton).toBeVisible();
  });

  test('should go to signup section when clicking the signup tab', () => {
    const signupButtons = screen.getAllByText('Sign Up');

    fireEvent.click(signupButtons[0]);

    expect(screen.getByText('Username')).toBeVisible();

    expect(screen.getByText('Password')).toBeVisible();

    expect(screen.getByText('Confirm Password')).toBeVisible();

    expect(screen.getByText('Country')).toBeVisible();

    expect(screen.getByText('Signup')).toBeVisible();
  });

  test('should submit to profile page when login', async () => {
    const loginButtons = screen.getAllByText('Login');

    const loginButton = loginButtons.find(
      (button) => button.className === 'button'
    ) as HTMLButtonElement;

    const usernameField = screen.getByLabelText('Username');

    const passwordField = screen.getByLabelText('Password');

    fireEvent.change(usernameField, { target: { value: 'test123' } });

    fireEvent.change(passwordField, { target: { value: 'testpassword123' } });

    await act(async () => {
      fireEvent.submit(loginButton);
    });
  });

  test('should match snapshot of auth page', () => {
    const container = render(<AuthPage />);

    expect(container).toMatchSnapshot();
  });
});
