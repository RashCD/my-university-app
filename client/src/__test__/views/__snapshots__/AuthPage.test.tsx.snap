// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for auth page should match snapshot of auth page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app"
      >
        <header
          class="appHeader"
        >
          <h1
            class="title"
          />
          <button
            class="button loginButton"
            type="button"
          >
            <span
              class="iconContainer"
              data-icon="true"
              style="--icon-src: url(user.svg); --icon-color: var(--white); --icon-size: 30px;"
            />
            Login
          </button>
        </header>
        <main
          class="app-main"
        >
          <div
            class="container"
          >
            <button
              class="button"
              type="button"
            >
              Login
            </button>
            <button
              class="button selectedLeft"
              type="button"
            >
              Sign Up
            </button>
            <div
              class="formContainer"
            >
              <form
                action="submit"
                class="form"
              >
                <div
                  class="inputContainer"
                >
                  <label
                    class="label"
                    for="username"
                  >
                    Username
                  </label>
                  <input
                    class="input"
                    id="username"
                    name="username"
                    placeholder="Enter Your Username"
                    type="text"
                  />
                </div>
                <div
                  class="inputContainer"
                >
                  <label
                    class="label"
                    for="password"
                  >
                    Password
                  </label>
                  <input
                    class="input"
                    id="password"
                    name="password"
                    placeholder="Enter Your Password"
                    type="password"
                  />
                </div>
                <button
                  class="button cta"
                  type="submit"
                >
                  Login
                </button>
              </form>
            </div>
          </div>
        </main>
      </div>
    </div>
    <div>
      <div
        class="app"
      >
        <header
          class="appHeader"
        >
          <h1
            class="title"
          />
          <button
            class="button loginButton"
            type="button"
          >
            <span
              class="iconContainer"
              data-icon="true"
              style="--icon-src: url(user.svg); --icon-color: var(--white); --icon-size: 30px;"
            />
            Login
          </button>
        </header>
        <main
          class="app-main"
        >
          <div
            class="container"
          >
            <button
              class="button"
              type="button"
            >
              Login
            </button>
            <button
              class="button selectedLeft"
              type="button"
            >
              Sign Up
            </button>
            <div
              class="formContainer"
            >
              <form
                action="submit"
                class="form"
              >
                <div
                  class="inputContainer"
                >
                  <label
                    class="label"
                    for="username"
                  >
                    Username
                  </label>
                  <input
                    class="input"
                    id="username"
                    name="username"
                    placeholder="Enter Your Username"
                    type="text"
                  />
                </div>
                <div
                  class="inputContainer"
                >
                  <label
                    class="label"
                    for="password"
                  >
                    Password
                  </label>
                  <input
                    class="input"
                    id="password"
                    name="password"
                    placeholder="Enter Your Password"
                    type="password"
                  />
                </div>
                <button
                  class="button cta"
                  type="submit"
                >
                  Login
                </button>
              </form>
            </div>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app"
    >
      <header
        class="appHeader"
      >
        <h1
          class="title"
        />
        <button
          class="button loginButton"
          type="button"
        >
          <span
            class="iconContainer"
            data-icon="true"
            style="--icon-src: url(user.svg); --icon-color: var(--white); --icon-size: 30px;"
          />
          Login
        </button>
      </header>
      <main
        class="app-main"
      >
        <div
          class="container"
        >
          <button
            class="button"
            type="button"
          >
            Login
          </button>
          <button
            class="button selectedLeft"
            type="button"
          >
            Sign Up
          </button>
          <div
            class="formContainer"
          >
            <form
              action="submit"
              class="form"
            >
              <div
                class="inputContainer"
              >
                <label
                  class="label"
                  for="username"
                >
                  Username
                </label>
                <input
                  class="input"
                  id="username"
                  name="username"
                  placeholder="Enter Your Username"
                  type="text"
                />
              </div>
              <div
                class="inputContainer"
              >
                <label
                  class="label"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="input"
                  id="password"
                  name="password"
                  placeholder="Enter Your Password"
                  type="password"
                />
              </div>
              <button
                class="button cta"
                type="submit"
              >
                Login
              </button>
            </form>
          </div>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
